src/Numerics/Mesh/DSS.jl

```diff
@@ -0,0 +1,318 @@
+
+# Fields
+ - `edgconn`: edge connectivity array
+ - `edgconnoff`: offsets for edge connectivity array
+ - `edgemap`: map to edge degrees of freedom: 
+    `edgemap[i, edgno, orient]` contains the element node index of 
+    the `i`th interior node on edge `edgno`, under orientation `orient`. 
+ - `Nemax`: # of relevant degrees of freedom per edge (other dof are marked as -1)
+ - `data`: data field of MPIStateArray
+ - `ivar`: variable # in the MPIStateArray
+ - `ex`: unique edge number
+ - `::Type{FT}`: Floating point type
+"""
+function dss_edge!(
+    edgconn,
+    edgconnoff,
+    edgemap,
+    Nemax,
+    data,
+    ivar,
+    ex,
+    ::Type{FT},
+) where {FT}
+    @inbounds st = edgconnoff[ex]
+    @inbounds nledg = Int((edgconnoff[ex + 1] - edgconnoff[ex]) / 3)
+
+    @inbounds for k in 1:Nemax
+        sume = -FT(0)
+        @inbounds for j in 1:nledg
+            ledg = edgconn[st + (j - 1) * 3]
+            lor = edgconn[st + (j - 1) * 3 + 1]
+            lelem = edgconn[st + (j - 1) * 3 + 2]
+            loc = edgemap[k, ledg, lor]
+            if loc ≠ -1
+                sume += data[loc, ivar, lelem]
+            end
+        end
+        @inbounds for j in 1:nledg
+            ledg = edgconn[st + (j - 1) * 3]
+            lor = edgconn[st + (j - 1) * 3 + 1]
+            lelem = edgconn[st + (j - 1) * 3 + 2]
+            loc = edgemap[k, ledg, lor]
+            if loc ≠ -1
+                data[loc, ivar, lelem] = sume
+            end
+        end
+    end
+end
+"""
+    dss_face!(fcconn, facemap, Nfmax, data, ivar, fx)
+
+This function computes the direct stiffness summation for 
+all degrees of freedom corresponding to face `fx`. dss_face!
+applies only to interior (non-vertex and non-edge) face nodes.
+
+# Fields
+ - `fcconn`: face connectivity array
+ - `facemap`: map to face degrees of freedom: `facemap[ij, fcno, orient]` 
+    contains the element node index of the `ij`th 
+    interior node on face `fcno` under orientation `orient`
+ - `Nfmax`: # of relevant degrees of freedom per face (other dof are marked as -1)
+ - `data`: data field of MPIStateArray
+ - `ivar`: variable # in the MPIStateArray
+ - `fx`: unique face number
+"""
+function dss_face!(fcconn, facemap, Nfmax, data, ivar, fx)
+    @inbounds lfc = fcconn[fx, 1]
+    @inbounds lel = fcconn[fx, 2]
+    @inbounds nabrlfc = fcconn[fx, 3]
+    @inbounds nabrlel = fcconn[fx, 4]
+    @inbounds ordr = fcconn[fx, 5]
+    ordr = ordr == 3 ? 2 : 1
+    # mesh orientation 3 is a flip along the horizontal edges.
+    # This is the only orientation currently support by the mesh generator
+    # see vertsortandorder in
+    # src/Numerics/Mesh/BrickMesh.jl                
+    @inbounds for j in 1:Nfmax
+        loc1 = facemap[j, lfc, ordr]
+        loc2 = facemap[j, nabrlfc, 1]
+        if loc1 ≠ -1 && loc2 ≠ -1
+            sumf = data[loc1, ivar, lel] + data[loc2, ivar, nabrlel]
+            data[loc1, ivar, lel] = sumf
+            data[loc2, ivar, nabrlel] = sumf
```

Comment:
Shouldn't this be weighted?

Reply:
Weighting should not be applied to DSS.

Reply:
For the postprocessing, weighting needs to be applied after dss. But this weight is just the inverse of duplication count (# of times the gldof is repeated), not quadrature weigths.

Reply:
Shouldn't it be
```suggestion
            sumf = (data[loc1, ivar, lel]*vgeo[loc1, _M, lel] + data[loc2, ivar, nabrlel]*vgeo[loc1, _M, nabrlel]) / (M[loc1, lel] + M[loc1, nabrlel])
            data[loc1, ivar, lel] = sumf
            data[loc2, ivar, nabrlel] = sumf
```

Reply:
Applying inverse mass matrix, etc. is a separate operation. They are unrelated.

Reply:
For performance, I think we will ultimately want to fuse the mass matrix operations with the dss and the velocity rotations (once those come into play).

But I agree that the dss is defined as just the gather/scatter operation without the mass matrix.

Reply:
(Though I've always thought it would make more sense for it to be an L2-project, but it's never defined this way)